tls/s2n_kem.c

```diff
@@ -23,44 +23,132 @@
 #include "utils/s2n_mem.h"
 #include "utils/s2n_safety.h"
 
+/* If S2N_NO_PQ is set or linked libcrypto doesn't support Kyber, bail on KEM calls.
+ * These should never be called.
+ */
+int s2n_kyber_kem_keypair_not_supported(IN const struct s2n_kem *kem, OUT uint8_t *pk, OUT uint8_t *sk)
+{
+    POSIX_BAIL(S2N_ERR_UNIMPLEMENTED);
+}
+int s2n_kyber_kem_enc_not_supported(IN const struct s2n_kem *kem, OUT uint8_t *ct, OUT uint8_t *ss,
+        IN const uint8_t *pk)
+{
+    POSIX_BAIL(S2N_ERR_UNIMPLEMENTED);
+}
+int s2n_kyber_kem_dec_not_supported(IN const struct s2n_kem *kem, OUT uint8_t *ss, IN const uint8_t *ct,
+        IN const uint8_t *sk)
+{
+    POSIX_BAIL(S2N_ERR_UNIMPLEMENTED);
+}
+
+#if defined(S2N_NO_PQ)
+/* If S2N_NO_PQ was defined at compile time, the PQ KEM code will have been entirely excluded
+ * from compilation. We define stubs of these functions here to error if they are called. */
+int s2n_kyber_512_r3_crypto_kem_keypair(IN const struct s2n_kem *kem, OUT uint8_t *pk, OUT uint8_t *sk)
+{
+    return s2n_kyber_kem_keypair_not_supported(kem, pk, sk);
+}
+int s2n_kyber_512_r3_crypto_kem_enc(IN const struct s2n_kem *kem, OUT uint8_t *ct, OUT uint8_t *ss,
+        IN const uint8_t *pk)
+{
+    return s2n_kyber_kem_enc_not_supported(kem, ct, ss, pk);
+}
+int s2n_kyber_512_r3_crypto_kem_dec(IN const struct s2n_kem *kem, OUT uint8_t *ss, IN const uint8_t *ct,
+        IN const uint8_t *sk)
+{
+    return s2n_kyber_kem_dec_not_supported(kem, ss, ct, sk);
+}
+#endif
+
 /* The KEM IDs and names come from https://tools.ietf.org/html/draft-campagna-tls-bike-sike-hybrid */
 
 const struct s2n_kem s2n_kyber_512_r3 = {
     .name = "kyber512r3",
+#if defined(S2N_LIBCRYPTO_SUPPORTS_KYBER)
+    .kem_nid = NID_KYBER512_R3,
+#else
+    .kem_nid = NID_undef,
+#endif
     .kem_extension_id = TLS_PQ_KEM_EXTENSION_ID_KYBER_512_R3,
     .public_key_length = S2N_KYBER_512_R3_PUBLIC_KEY_BYTES,
     .private_key_length = S2N_KYBER_512_R3_SECRET_KEY_BYTES,
     .shared_secret_key_length = S2N_KYBER_512_R3_SHARED_SECRET_BYTES,
     .ciphertext_length = S2N_KYBER_512_R3_CIPHERTEXT_BYTES,
-#if defined(S2N_LIBCRYPTO_SUPPORTS_KYBER)
-    .generate_keypair = &s2n_kyber_512_evp_generate_keypair,
-    .encapsulate = &s2n_kyber_512_evp_encapsulate,
-    .decapsulate = &s2n_kyber_512_evp_decapsulate,
+#if defined(S2N_LIBCRYPTO_SUPPORTS_KYBER) && !defined(S2N_NO_PQ)
+    .generate_keypair = &s2n_kyber_evp_generate_keypair,
+    .encapsulate = &s2n_kyber_evp_encapsulate,
+    .decapsulate = &s2n_kyber_evp_decapsulate,
 #else
     .generate_keypair = &s2n_kyber_512_r3_crypto_kem_keypair,
     .encapsulate = &s2n_kyber_512_r3_crypto_kem_enc,
     .decapsulate = &s2n_kyber_512_r3_crypto_kem_dec,
 #endif
```

Comment:
I assume all the conditional compilation here will be addressed by https://github.com/aws/s2n-tls/issues/4100?

Reply:
not all of it. we'll still need to do conditional compilation to point this struct to the interned Kyber512 implementation in the case where PQ is enabled but s2n-tls isn't built against AWS-LC. once we remove the interned Kyber implementation later this quarter we'll be able to remove that conditional compilation and do it all at runtime.

Reply:
Ideally, the only conditional compilation in our code should be to work around symbols that don't exist.

Since there are no direct calls to libcrypto methods here (except maybe NID_KYBER512_R3?) we should be able to remove the conditional compilation. If both symbols exist, we should be able to choose between them at runtime. That'll likely mean moving the conditional compilation further down the callstack to get it closer to the missing symbol.

Reply:
that makes sense. I'll remove the conditional compilation for the function pointers in this struct as part of PR 4101. the `NID_KYBER512_R3` symbol only exists in AWS-LC, so we'll need to retain that portion of conditional compilation.

Reply:
ah, i see. i'll alias that under the `S2N_` namespace.